ports/esp32/machine_can.c

```diff
@@ -0,0 +1,673 @@
+STATIC mp_obj_t machine_hw_can_any(mp_obj_t self_in){
+    can_status_info_t status = _machine_hw_can_get_status();
+    return mp_obj_new_bool((status.msgs_to_rx) > 0);
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_1(machine_hw_can_any_obj, machine_hw_can_any);
+
+// Get the state of the controller
+STATIC mp_obj_t machine_hw_can_state(mp_obj_t self_in){
+    can_status_info_t status = _machine_hw_can_get_status();
+    return mp_obj_new_int(status.state);
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_1(machine_hw_can_state_obj, machine_hw_can_state);
+
+// Get info about error states and TX/RX buffers
+STATIC mp_obj_t machine_hw_can_info(size_t n_args, const mp_obj_t *args){
+    machine_can_obj_t *self = MP_OBJ_TO_PTR(args[0]);
+    mp_obj_list_t *list;
+    if (n_args == 1){
+        list = MP_OBJ_TO_PTR(mp_obj_new_list(8, NULL));
+    }else{
+        if (!mp_obj_is_type(args[1], &mp_type_list)){
+            mp_raise_TypeError(NULL);
+        }
+        list = MP_OBJ_TO_PTR(args[1]);
+        if (list->len < 8){
+            mp_raise_ValueError(NULL);
+        }
+    }
+    can_status_info_t status = _machine_hw_can_get_status();
+    list->items[0] = MP_OBJ_NEW_SMALL_INT(status.tx_error_counter);
+    list->items[1] = MP_OBJ_NEW_SMALL_INT(status.rx_error_counter);
+    list->items[2] = MP_OBJ_NEW_SMALL_INT(self->num_error_warning);
+    list->items[3] = MP_OBJ_NEW_SMALL_INT(self->num_error_passive);
+    list->items[4] = MP_OBJ_NEW_SMALL_INT(self->num_bus_off);
+    list->items[5] = MP_OBJ_NEW_SMALL_INT(status.msgs_to_tx);
+    list->items[6] = MP_OBJ_NEW_SMALL_INT(status.msgs_to_rx);
+    list->items[7] = mp_const_none;
+    return MP_OBJ_FROM_PTR(list);
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_VAR_BETWEEN(machine_hw_can_info_obj, 1, 2, machine_hw_can_info);
+
+// Get Alert info
+STATIC mp_obj_t machine_hw_can_alert(mp_obj_t self_in){
+    uint32_t alerts;
+    uint32_t status = can_read_alerts(&alerts, 0);
+    if (status != ESP_OK){
+        mp_raise_OSError(-status);
+    }
+    return mp_obj_new_int(alerts);
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_1(machine_hw_can_alert_obj, machine_hw_can_alert);
+
+// Clear TX Queue
+STATIC mp_obj_t machine_hw_can_clear_tx_queue(mp_obj_t self_in){
+    return mp_obj_new_bool(can_clear_transmit_queue() == ESP_OK);
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_1(machine_hw_can_clear_tx_queue_obj, machine_hw_can_clear_tx_queue);
+
+// Clear RX Queue
+STATIC mp_obj_t machine_hw_can_clear_rx_queue(mp_obj_t self_in){
+    return mp_obj_new_bool(can_clear_receive_queue() == ESP_OK);
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_1(machine_hw_can_clear_rx_queue_obj, machine_hw_can_clear_rx_queue);
+
+// send([data], id, *)
+STATIC mp_obj_t machine_hw_can_send(size_t n_args, const mp_obj_t *pos_args, mp_map_t *kw_args){
+    enum{
+        ARG_data,
+        ARG_id,
+        ARG_timeout,
+        ARG_rtr,
+        ARG_self
+    };
+    static const mp_arg_t allowed_args[] = {
+        {MP_QSTR_data, MP_ARG_REQUIRED | MP_ARG_OBJ, {.u_obj = MP_OBJ_NULL}},
```

Comment:
No default required for required params.

Reply:
Why was this marked as resolved?